src/lib/cpd/CpdEngine.ts

```diff
@@ -0,0 +1,215 @@
+import {Logger, SfdxError} from '@salesforce/core';
+import {xml2js} from 'xml-js';
+import checksum = require('checksum')
+import {Controller} from '../../Controller';
+import {Catalog, Rule, RuleGroup, RuleResult, RuleTarget, RuleViolation, TargetPattern} from '../../types';
+import {AbstractRuleEngine} from '../services/RuleEngine';
+import {Config} from '../util/Config';
+import {ENGINE, Severity} from '../../Constants';
+import {PmdCatalogWrapper} from '../pmd/PmdCatalogWrapper';
+import * as engineUtils from '../util/CommonEngineUtils';
+import CpdWrapper from './CpdWrapper';
+
+
+export class CpdEngine extends AbstractRuleEngine {
+
+	public static ENGINE_ENUM: ENGINE = ENGINE.CPD;
+	public static ENGINE_NAME: string = ENGINE.CPD.valueOf();
+
+	private logger: Logger;
+	private config: Config;
+	private initialized: boolean;
+	private cpdCatalog: Catalog;
+	private languageToPathMap;
+	protected pmdCatalogWrapper: PmdCatalogWrapper;
+
+	public getName(): string {
+		return CpdEngine.ENGINE_NAME;
+	}
+
+	public getTargetPatterns(): Promise<TargetPattern[]> {
+		return this.config.getTargetPatterns(ENGINE.CPD);
+	}
+
+	public getCatalog(): Promise<Catalog>{
+		return Promise.resolve(this.cpdCatalog);
+	}
+
+	/* eslint-disable-next-line no-unused-vars, @typescript-eslint/no-unused-vars */
+	public shouldEngineRun(ruleGroups: RuleGroup[], rules: Rule[], target: RuleTarget[], engineOptions: Map<string, string>): boolean {
+		return true;
+	}
+	
+	// eslint-disable-next-line @typescript-eslint/no-unused-vars
+	public async run(ruleGroups: RuleGroup[], rules: Rule[], targets: RuleTarget[], engineOptions: Map<string, string>): Promise<RuleResult[]> {
+		try {
+			const results = []
+			// cpd only allows one language to be speificed at a time, so we must run it on each (apex and vf)
+			for (const language of this.cpdCatalog.rules[0].languages) {
+				const targetPaths: string[] = [];
+				for (const target of targets) {
```

Comment:
Having target path building in its own method would help with readability as well.